Api/MyMods.cs

```diff
@@ -51,13 +57,16 @@ class MyMods {
             foreach (Panel p in modpacksList) {
                 CheckBox chb = (CheckBox)p.GetChildAtPoint(Config.MyModsChbPoint);
                 string modpackname = chb.Text.Replace(Config.dirtyPadding, "");
-                if (chb.Checked && !Config.IsPatched(modpackname)) {
-                    toPatch.Add(chb);
-                    if ((string)p.Tag != Config.GetCurrentBuild()) {
-                        oldModpacks.Add(modpackname);
+                if (chb.Checked) {
```

Comment:
This change breaks how the application currently works. The code you removed checks each row, if the light is red and the checkbox is checked, the manager will patch the game (your new code does this). However, if the light is green and the checkbox is unchecked the mod manager should unpatch it (your new code swaps this, unpatching if it is checked).

I don't think it's a good idea to swap the controls on the users like this, people will get annoyed because they are already used to how it works (it's been like this since initial release). If anything is checked, when the patch button is clicked it adds those mods to the game, and makes sure any unchecked mods are removed from the game. Changing this would required making changes elsewhere too for QoL improvements. Currently when the mod manager loads, it auto checks any modpack which is enabled. With this change, a user simply trying to add one mod to their game would check a new box and then suddenly every patched mod would be uninstalled, leaving only the new one. This also creates a toggling issue, where clicking the button twice because the app is slow could cause the manager to install and then immediately uninstall all mods.

Lets reset this back to how it was

Reply:
Hrmm. So that behavior was intentional. That's an...intriguing design decision. I modified it because, once I encountered it, I intuitively believed it to be a bug.

> if the light is red and the checkbox is checked, the manager will patch the game 
> if the light is green and the checkbox is unchecked the mod manager should unpatch it

This makes no sense to me. For a first time user, the Patch/Unpatch button appears to apply its Patch/Unpatch functionality to the mods that are currently checked, regardless of if they active or inactive mods. It's a process that's applied on a **per selection** basis.

If I open the manager to a fresh MyMods page, to a blank and completely unchecked screen, I intuitively wouldn't expect the Patch/Unpatch button to do anything. But since you've inverted the behavior for already patched mods, it behaves oppositely from unpatched mods with no visual cues to tell the user it behaves differently--making it apply an operations on a blank page with no selections.

Even the error message you provide lends itself towards the behavior I'm expecting. See the output that's generated when I revert this change:
![Capture](https://user-images.githubusercontent.com/49787409/77380250-0b71b000-6d51-11ea-82f6-7dfe01c4058b.PNG)

From what you're telling me, the above is expected behavior.
I think if you want Patch/Unpatch to behave differently for patched vs unpatched mods, the UX for one of the two operations needs to change. Identical UX should result in identical behavior.

> Currently when the mod manager loads, it auto checks any modpack which is enabled.
I see why you were doing this for convenience, but don't you think the new Select Enabled bulk checkbox does this now?

> This also creates a toggling issue, where clicking the button twice because the app is slow could cause the manager to install and then immediately uninstall all mods.
We should probably block input to that button until the operation is complete, regardless of how the checkboxes work.

Reply:
Yes this is intended behavior. We could perhaps change that popup you screenshot to instead read "You did not MAKE any changes", instead of "select" to be more clear. But this is how the mod manager has worked since inception and I haven't received any complaints about it. I think it is convenient in the sense that anything that is checked is going to be patched to the game, and anything that is unchecked is not going to be patched to the game, after pressing the button, regardless of the current state of your mods. Also, given the same UI state that button _does_ perform the same action each time. You aren't taking into account the state of the red/green button, which is a part of the UI.

As for the toggle issue, I was more referring to someone spamming the button while the process finishes, returning control to the UI while the user is spamming, causing them to click the button again and unpatch everything.

Really, either control scheme works. So I think it's best that we don't make a change that isn't entirely necessary, especially since the userbase is already used to this one.

Reply:
>after pressing the button, any mod that is checked ends up installed in the game, and any mod that is unchecked ends up as not installed in the game, regardless of the current state of your mods installations.

I checked back in previous versions, as the version I'd been using for so long doesn't behave this way. Turns out I'd changed this on my personal branch a long time ago and forgotten about it.
I can understand this line of thinking--you're aligning the checkbox selection with installation. Checked means installed, unchecked means not installed, and patch unpatch will _align_ the modpacks to that state.
As opposed to my line of thinking, where the selection applies not to that state, but directly to the patch/unpatch functionality. Whatever is selected will apply patch/unpatch, regardless of the state of the mod.

I'm genuinely amazed that you haven't gotten a complaint up to this point. While your approach is most certainly valid, The current UI implicitly lends itself towards my implementation, between the error message and the column name. I strongly recommend making some changes to make it obvious.
Instead of calling the column name 'Select', it could be called 'Mark for install/uninstall' as the user isn't selecting a mod to run patch/unpatch, they're marking what state the pack should be in. Then the button could be called 'Align' to align the state of the mods with what's marked. Or 'Match Selections'. I still don't love this but at least it's then clear. I don't want users to encounter this and believe it to be a bug rather than an intentional design decision.

Since both approaches are similar, we could add another checkbox to the top right of the page to flip between the two types of control modes. If it's intuitive to me, it will likely be intuitive to many others.

> As for the toggle issue, I was more referring to someone spamming the button while the process finishes, returning control to the UI while the user is spamming, causing them to click the button again and unpatch everything.

...Which is exactly why I said:

> We should probably block input to that button until the operation is complete, regardless of how the checkboxes work.

This should stop people from spamming patch/unpatch operations.

I'll revert this. (Though you can bet I'm keeping it in my personal build ;P )

Reply:
> this is how the mod manager has worked since inception and I haven't received any complaints about it.

Not believing this to be possible, I checked in older versions and found that I'd already made this change in my _personal_ version of 0.5, and forgotten that I'd done so. All you really have to do is delete the one !.

That said, I understand your line of thinking--you're making the checkbox selection 1-to-1 with the state of being installed vs not installed, regardless of the state the mod is already in. This approach is certainly valid, however I think your current UI _greatly_ lends itself implicitly towards my implementation, between the error message and the column name. I strongly recommend that you make tweaks to make your control scheme more obvious. Without doing so users will absolutely make the same mistake, and believe it to be a bug rather than in intentional design decision. Users thinking opposite you probably fiddled with it enough to get it to work, and begrudgingly accepted it rather than report it.

You could change the column name to 'Mark as installed'. Then change the Patch/Unpatch text could change to 'Align' or 'Match Selections' to make the mods match the checkboxes. I don't love this, but makes your control scheme apparent. I think the difficulty in making that approach clearer is a testament to the cumbersome nature it has.

Additionally, we could add a checkbox to flip between the two types of control schemes. It would be easy to implement.

> As for the toggle issue, I was more referring to someone spamming the button while the process finishes, returning control to the UI while the user is spamming, causing them to click the button again and unpatch everything.

...that's why I said:

> We should probably block input to that button until the operation is complete, regardless of how the checkboxes work.

I've reverted the checkbox logic. But you better believe I'll be keeping it in my personal branch ;P